File: test-7-agent-system.js
#!/usr/bin/env node
/**
 * Test Script for 7-Agent Medical Evidence Synthesis System
 * Tests the 10th query from openevidence.md using Node.js
 */

const { spawn } = require('child_process');
const fs = require('fs');
const path = require('path');

// The 10th query from openevidence.md
const testQuery = `In severe asthma with eosinophilic phenotype and frequent exacerbations despite high‑dose ICS/LABA, how should clinicians choose between mepolizumab, benralizumab, dupilumab, and tezepelumab based on biomarker profiles (eosinophils, FeNO, IgE), comorbidities, and head‑to‑head or indirect comparative evidence?`;

console.log('🚀 Testing 7-Agent Medical Evidence Synthesis System');
console.log('=' .repeat(80));
console.log(`📋 Test Query: ${testQuery}`);
console.log('=' .repeat(80));
console.log(`⏰ Start Time: ${new Date().toLocaleString()}`);
console.log();

// Test via API endpoint
async function testViaAPI() {
    console.log('🌐 Testing via Next.js API endpoint...');
    
    try {
        // Start the Next.js development server
        console.log('🔧 Starting Next.js development server...');
        
        const server = spawn('npm', ['run', 'dev'], {
            stdio: ['pipe', 'pipe', 'pipe'],
            shell: true
        });
        
        let serverReady = false;
        let serverOutput = '';
        
        // Wait for server to be ready
        server.stdout.on('data', (data) => {
            const output = data.toString();
            serverOutput += output;
            console.log('📡 Server:', output.trim());
            
            if (output.includes('Ready') || output.includes('localhost:3000')) {
                serverReady = true;
            }
        });
        
        server.stderr.on('data', (data) => {
            console.log('⚠️ Server Error:', data.toString().trim());
        });
        
        // Wait for server to be ready (max 30 seconds)
        let waitTime = 0;
        while (!serverReady && waitTime < 30000) {
            await new Promise(resolve => setTimeout(resolve, 1000));
            waitTime += 1000;
            if (waitTime % 5000 === 0) {
                console.log(`⏳ Waiting for server... (${waitTime/1000}s)`);
            }
        }
        
        if (!serverReady) {
            console.log('❌ Server failed to start within 30 seconds');
            server.kill();
            return false;
        }
        
        console.log('✅ Server is ready! Making API request...');
        
        // Make API request
        const fetch = (await import('node-fetch')).default;
        
        const response = await fetch('http://localhost:3000/api/chat', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
            },
            body: JSON.stringify({
                messages: [
                    { role: 'user', content: testQuery }
                ],
                mode: 'doctor',
                sessionId: `test_session_${Date.now()}`
            })
        });
        
        console.log(`📊 Response Status: ${response.status}`);
        console.log(`📊 Response Headers:`, Object.fromEntries(response.headers.entries()));
        
        if (!response.ok) {
            const errorText = await response.text();
            console.log('❌ API Error:', errorText);
            server.kill();
            return false;
        }
        
        // Handle streaming response
        console.log('\n🎯 Processing streaming response...');
        console.log('-' .repeat(40));
        
        let fullResponse = '';
        let citationsCount = 0;
        let sourcesCount = 0;
        let cost = 0;
        let latency = 0;
        let groundingScore = 0;
        
        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        
        while (true) {
            const { done, value } = await reader.read();
            if (done) break;
            
            const chunk = decoder.decode(value);
            const lines = chunk.split('\n');
            
            for (const line of lines) {
                if (line.startsWith('data: ')) {
                    const data = line.slice(6);
                    if (data === '[DONE]') {
                        console.log('\n✅ Stream completed');
                        break;
                    }
                    
                    try {
                        const parsed = JSON.parse(data);
                        
                        if (parsed.content) {
                            process.stdout.write(parsed.content);
                            fullResponse += parsed.content;
                        }
                        
                        if (parsed.citations) {
                            citationsCount = parsed.citations.length;
                        }
                        
                        if (parsed.sources_count !== undefined) {
                            sourcesCount = parsed.sources_count;
                            cost = parsed.cost || 0;
                            latency = parsed.latency || 0;
                            groundingScore = parsed.grounding_score || 0;
                        }
                        
                        if (parsed.warning) {
                            console.log(`\n⚠️ Warning: ${parsed.warning}`);
                        }
                        
                    } catch (parseError) {
                        // Ignore parse errors for partial chunks
                    }
                }
            }
        }
        
        console.log('\n\n' + '=' .repeat(80));
        console.log('📊 FINAL RESULTS');
        console.log('=' .repeat(80));
        console.log(`✅ Response Length: ${fullResponse.length} characters`);
        console.log(`📚 Citations Count: ${citationsCount}`);
        console.log(`🔍 Sources Count: ${sourcesCount}`);
        console.log(`⏱️  Total Latency: ${latency}ms`);
        console.log(`💰 Total Cost: $${cost.toFixed(4)}`);
        console.log(`🎯 Grounding Score: ${Math.round(groundingScore * 100)}%`);
        
        console.log('\n🎉 TEST COMPLETED SUCCESSFULLY!');
        console.log('=' .repeat(80));
        
        // Kill the server
        server.kill();
        return true;
        
    } catch (error) {
        console.error('❌ Test failed:', error);
        return false;
    }
}

// Test via direct TypeScript compilation and execution
async function testViaTypeScript() {
    console.log('🔧 Testing via direct TypeScript execution...');
    
    try {
        // Check if we can compile and run TypeScript directly
        const tsNode = spawn('npx', ['ts-node', '--esm', '-e', `
            import { MedicalEvidenceOrchestrator } from './lib/agents/medical-evidence-orchestrator.js';
            
            const config = {
                google_ai_api_key: process.env.GEMINI_API_KEY || '',
                ncbi_api_key: process.env.NCBI_API_KEY || '',
                tavily_api_key: process.env.TAVILY_API_KEY || ''
            };
            
            const orchestrator = new MedicalEvidenceOrchestrator(config);
            const query = \`${testQuery}\`;
            
            console.log('🚀 Testing orchestrator directly...');
            
            orchestrator.processQuery(query, 'test_session_direct')
                .then(response => {
                    console.log('✅ Direct test successful!');
                    console.log('Response:', JSON.stringify(response, null, 2));
                })
                .catch(error => {
                    console.error('❌ Direct test failed:', error);
                    process.exit(1);
                });
        `], {
            stdio: 'inherit',
            shell: true,
            env: { ...process.env }
        });
        
        return new Promise((resolve) => {
            tsNode.on('close', (code) => {
                resolve(code === 0);
            });
        });
        
    } catch (error) {
        console.error('❌ TypeScript test failed:', error);
        return false;
    }
}

async function main() {
    console.log('🧪 7-Agent Medical Evidence Synthesis System Test');
    console.log('Testing with Query #10 from openevidence.md');
    console.log();
    
    // Check if node-fetch is available
    try {
        await import('node-fetch');
    } catch (error) {
        console.log('📦 Installing node-fetch...');
        const install = spawn('npm', ['install', 'node-fetch'], { stdio: 'inherit', shell: true });
        await new Promise((resolve) => {
            install.on('close', resolve);
        });
    }
    
    // Try API test first
    console.log('🎯 Method 1: Testing via Next.js API...');
    const apiSuccess = await testViaAPI();
    
    if (apiSuccess) {
        console.log('🎉 API test completed successfully!');
        return true;
    }
    
    console.log('\n🎯 Method 2: Testing via direct TypeScript...');
    const tsSuccess = await testViaTypeScript();
    
    if (tsSuccess) {
        console.log('🎉 TypeScript test completed successfully!');
        return true;
    }
    
    console.log('❌ Both test methods failed');
    return false;
}

// Run the test
main()
    .then(success => {
        process.exit(success ? 0 : 1);
    })
    .catch(error => {
        console.error('❌ Unexpected error:', error);
        process.exit(1);
    });

File: test-7-agent-system.py
#!/usr/bin/env python3
"""
Test Script for 7-Agent Medical Evidence Synthesis System
Tests the 10th query from openevidence.md
"""

import asyncio
import json
import os
import sys
from datetime import datetime
from dotenv import load_dotenv

# Load environment variables from .env.local
load_dotenv('.env.local')

# Add the lib directory to Python path
sys.path.append(os.path.join(os.path.dirname(__file__), 'lib'))

async def test_7_agent_system():
    """Test the 7-agent system with the 10th query from openevidence.md"""
    
    # The 10th query from openevidence.md
    test_query = """In severe asthma with eosinophilic phenotype and frequent exacerbations despite high‑dose ICS/LABA, how should clinicians choose between mepolizumab, benralizumab, dupilumab, and tezepelumab based on biomarker profiles (eosinophils, FeNO, IgE), comorbidities, and head‑to‑head or indirect comparative evidence?"""
    
    print("🚀 Testing 7-Agent Medical Evidence Synthesis System")
    print("=" * 80)
    print(f"📋 Test Query: {test_query}")
    print("=" * 80)
    print(f"⏰ Start Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print()
    
    try:
        # Import the orchestrator
        from lib.agents.medical_evidence_orchestrator import MedicalEvidenceOrchestrator
        
        # Initialize with environment variables
        config = {
            'google_ai_api_key': os.getenv('GEMINI_API_KEY'),
            'ncbi_api_key': os.getenv('NCBI_API_KEY', ''),
            'tavily_api_key': os.getenv('TAVILY_API_KEY', '')
        }
        
        # Validate configuration
        if not config['google_ai_api_key']:
            raise ValueError("GEMINI_API_KEY environment variable is required")
        
        print("🔧 Initializing 7-Agent Medical Evidence Orchestrator...")
        orchestrator = MedicalEvidenceOrchestrator(config)
        
        print("✅ Orchestrator initialized successfully")
        print()
        
        # Process the query
        print("🎯 Processing query through 7-agent system...")
        session_id = f"test_session_{int(datetime.now().timestamp())}"
        
        response = await orchestrator.processQuery(test_query, session_id)
        
        print("\n" + "=" * 80)
        print("📊 FINAL RESULTS")
        print("=" * 80)
        
        # Display results
        print(f"✅ Synthesis Length: {len(response['synthesis'])} characters")
        print(f"📚 Citations Count: {len(response['citations'])}")
        print(f"🔍 Sources Count: {response['metadata']['sources_count']}")
        print(f"⏱️  Total Latency: {response['metadata']['latency_total_ms']}ms")
        print(f"💰 Total Cost: ${response['metadata']['cost_total_usd']:.4f}")
        print(f"🎯 Grounding Score: {response['metadata']['grounding_score']:.2f}")
        print(f"🤖 Model Used: {response['metadata']['model_used']}")
        print(f"🆔 Trace ID: {response['metadata']['trace_id']}")
        
        if response['metadata']['hallucination_detected']:
            print("⚠️  Hallucination Detected: YES")
        else:
            print("✅ Hallucination Detected: NO")
        
        print("\n" + "-" * 40)
        print("📝 SYNTHESIS PREVIEW (First 500 chars):")
        print("-" * 40)
        print(response['synthesis'][:500] + "..." if len(response['synthesis']) > 500 else response['synthesis'])
        
        print("\n" + "-" * 40)
        print("📚 CITATIONS:")
        print("-" * 40)
        for i, citation in enumerate(response['citations'][:5], 1):  # Show first 5 citations
            print(f"{i}. {citation['title'][:80]}...")
            print(f"   Source: {citation['source']} | ID: {citation['id']}")
            if citation.get('url'):
                print(f"   URL: {citation['url']}")
            print()
        
        if len(response['citations']) > 5:
            print(f"... and {len(response['citations']) - 5} more citations")
        
        if response.get('warning'):
            print("\n⚠️  WARNING:")
            print(response['warning'])
        
        print("\n" + "=" * 80)
        print("🎉 TEST COMPLETED SUCCESSFULLY!")
        print("=" * 80)
        
        return True
        
    except ImportError as e:
        print(f"❌ Import Error: {e}")
        print("💡 Make sure the 7-agent system is properly installed")
        return False
        
    except Exception as e:
        print(f"❌ Test Failed: {e}")
        print(f"📋 Error Type: {type(e).__name__}")
        import traceback
        print("📋 Full Traceback:")
        traceback.print_exc()
        return False

def main():
    """Main function to run the test"""
    print("🧪 7-Agent Medical Evidence Synthesis System Test")
    print("Testing with Query #10 from openevidence.md")
    print()
    
    # Check environment variables
    required_vars = ['GEMINI_API_KEY']
    missing_vars = [var for var in required_vars if not os.getenv(var)]
    
    if missing_vars:
        print("❌ Missing required environment variables:")
        for var in missing_vars:
            print(f"   - {var}")
        print("\n💡 Please set these variables in your .env.local file")
        return False
    
    # Run the async test
    try:
        result = asyncio.run(test_7_agent_system())
        return result
    except KeyboardInterrupt:
        print("\n⏹️  Test interrupted by user")
        return False
    except Exception as e:
        print(f"\n❌ Unexpected error: {e}")
        return False

if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)